Implementation of AoE Blacklist for Tinker's Construct Tools

AlexanderChenhall opened the pull request:
This feature would enhance control over tool behavior and can prevent accidental destruction of important blocks, think accidentally breaking a wall of chests with a lumber axe, or destroying your Smeltery Controller when breaking Seared brick blocks.

Key Changes:
Created new class under tconstruct.library.util to handle loading and checking exclusions for AoE tool effects.

Lists are determined by a config (preferably pre-set with desired defaults).
Lists are separated into exclusions lists for individual tools, so a separate blacklist can exist for the hammer, separate from the excavator or lumberaxe.

Updated the AOEHarvestTool class to add a check before breaking blocks via AoE, of course keeps original functionality.

Did the whole gauntlet of gradle tests, no issues. Of course i may have missed something so feel free to have a look and/or reject if there was an oversight or you feel it wouldn't fit within the mod/pack as present.


Example of the config file:
```
# Configuration file

aoe_exclusions {
    # Block IDs that should not be broken by tool.excavator's AOE effect [default: ]
    S:tool.excavatorExclusions <
        examplemod:exampleblock
     >

    # Block IDs that should not be broken by tool.hammer's AOE effect [default: ]
    S:tool.hammerExclusions <
        minecraft:furnace
        minecraft:ender_chest
        TConstruct:ToolForgeBlock
        catwalks:scaffold
     >

    # Block IDs that should not be broken by tool.lumberaxe's AOE effect [default: ]
    S:tool.lumberaxeExclusions <
        minecraft:chest
     >
}

```

OneEyeMaker commented on README.md:
Please do not modify the README file without good reason. If you want to indicate your intent - please do so in the description of the PR.

AlexanderChenhall replied:
My bad, i did not intend to change the README at all, or at the very least, pulling in the one i changed was a mistake.

OneEyeMaker reviewed (changes requested):
Good start.

Please make tool names consistent and clean up code.

Also I suggest to extend functionality of the blacklist to allow blacklisting blocks only with specific metadata.

AlexanderChenhall commented on src/main/java/tconstruct/library/util/AoEExclusionList.java:
```diff
@@ -0,0 +1,55 @@
+
+public class AoEExclusionList {
+
+    private static final Map<String, String[]> toolExclusionLists = new HashMap<>();
```
Done.

AlexanderChenhall commented on src/main/java/tconstruct/library/util/AoEExclusionList.java:
```diff
@@ -0,0 +1,55 @@
+                // Try without "tool." prefix if it's present
+                exclusions = toolExclusionLists.get(tool.substring(5) + "Exclusions");
+            }
+        }
```
Done

AlexanderChenhall commented on src/main/java/tconstruct/library/tools/AOEHarvestTool.java:
```diff
@@ -18,9 +20,14 @@ public AOEHarvestTool(int baseDamage, int breakRadius, int breakDepth) {
+        String toolName = "tool." + getToolName().toLowerCase();
         Block block = player.worldObj.getBlock(x, y, z);
         int meta = player.worldObj.getBlockMetadata(x, y, z);
         if (block == null || !isEffective(block, meta) || !stack.hasTagCompound())
```
This is actually the intent. The hammer should still be able to break the block you're looking at, as users will very likely only use a hammer as their mining tool at these stages of the game.
For example, if you still wanted to break an iron chest, the intent is to still let you do that, and have it not break the chests left and right of it too on accident, as is unfortunately common.

AlexanderChenhall commented on src/main/java/tconstruct/library/tools/AOEHarvestTool.java:
```diff
@@ -64,8 +71,13 @@ public boolean onBlockStartBreak(ItemStack stack, int x, int y, int z, EntityPla
-                if (!super.onBlockStartBreak(stack, xPos, yPos, zPos, player))
-                    breakExtraBlock(player.worldObj, xPos, yPos, zPos, sideHit, player, x, y, z);
+                Block targetBlock = player.worldObj.getBlock(xPos, yPos, zPos);
+                String blockId = Block.blockRegistry.getNameForObject(targetBlock);
```
String Variable removed.

AlexanderChenhall commented on src/main/java/tconstruct/library/tools/AOEHarvestTool.java:
```diff
@@ -64,8 +71,13 @@ public boolean onBlockStartBreak(ItemStack stack, int x, int y, int z, EntityPla
+                Block targetBlock = player.worldObj.getBlock(xPos, yPos, zPos);
+                String blockId = Block.blockRegistry.getNameForObject(targetBlock);
+
+                if (!AoEExclusionList.isBlockExcluded(getAOEToolName(), targetBlock)) {
```
Passing down toolName variable from earlier instead of calling, done.

AlexanderChenhall commented on src/main/java/tconstruct/library/tools/AOEHarvestTool.java:
```diff
@@ -18,9 +20,14 @@ public AOEHarvestTool(int baseDamage, int breakRadius, int breakDepth) {
     public boolean onBlockStartBreak(ItemStack stack, int x, int y, int z, EntityPlayer player) {
-        // only effective materials matter. We don't want to aoe when beraking dirt with a hammer.
+        // only effective materials matter. We don't want to aoe when breaking dirt with a hammer.
+        String toolName = "tool." + getToolName().toLowerCase();
```
changed it to simply caching the tool name, this will get passed down instead of calling on the getAOEToolName later on. LMK if that's what you meant

AlexanderChenhall commented:
I'm back to say I've (hopefully) addressed all the issues you raised.
Additionally, there is now support for metadata. So you could for example, blacklist catwalks:scaffold:1 and not catwalks:scaffold

OneEyeMaker reviewed (approved):
Good job!

In current state this feature works completely fine (tested). As I said before, it would be nice to have option to blacklist blocks with specific metadata.

Dream-Master commented:
@AlexanderChenhall can you make a config in the Modpack repo so we can use this feature. Maybe ask around which blocks should be set as standard

AlexanderChenhall commented:
Before anything there is one more commit.
For some reason my IDE automatically changed
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

to

import java.util.*

Should be fixed now.

AlexanderChenhall commented:
> @AlexanderChenhall can you make a config in the Modpack repo so we can use this feature. Maybe ask around which blocks should be set as standard

Yes sir 🫡
